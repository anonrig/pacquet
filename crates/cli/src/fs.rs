use std::{io, os, path::Path};

#[cfg(unix)]
pub fn symlink_dir<P: AsRef<Path>>(original: P, link: P) -> io::Result<()> {
    os::unix::fs::symlink(original, link)
}

#[cfg(windows)]
pub fn symlink_dir<P: AsRef<Path>>(original: P, link: P) -> io::Result<()> {
    os::windows::fs::symlink_dir(original, link)
}

#[cfg(test)]
pub fn get_filenames_in_folder(path: &Path) -> Vec<String> {
    let mut files = std::fs::read_dir(path)
        .unwrap()
        .map(|e| e.unwrap().file_name().to_string_lossy().to_string())
        .collect::<Vec<_>>();

    files.sort();
    files
}

#[cfg(test)]
pub fn get_all_folders(root: &std::path::PathBuf) -> Vec<String> {
    let mut files = Vec::new();
    for entry in walkdir::WalkDir::new(root) {
        let entry = entry.unwrap();
        let entry_path = entry.path().to_path_buf();
        if entry.file_type().is_dir() || entry.file_type().is_symlink() {
            // We need this mutation to ensure that both Unix and Windows paths resolves the same.
            // TODO: Find a better way to do this?
            let simple_path = entry_path
                .strip_prefix(root)
                .unwrap()
                .components()
                .map(|c| c.as_os_str().to_string_lossy().to_string())
                .collect::<Vec<_>>()
                .join("/");

            if !simple_path.is_empty() {
                files.push(simple_path.as_os_string());
            }
        }
    }
    files.sort();
    files
}


