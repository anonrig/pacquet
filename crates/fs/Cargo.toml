[package]
name                 = "pacquet-fs"
description          = "Filesystem utility functions used by pacquet"
version              = "0.0.1"
publish              = false
authors.workspace    = true
edition.workspace    = true
homepage.workspace   = true
keywords.workspace   = true
license.workspace    = true
repository.workspace = true

[dependencies]
pacquet-task-queue = { workspace = true }

derive_more = { workspace = true }
miette      = { workspace = true }

[target.'cfg(windows)'.dependencies]
junction = { workspace = true }
