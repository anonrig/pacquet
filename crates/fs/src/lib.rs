mod ensure_file;
mod symlink_dir;

pub use ensure_file::*;
pub use symlink_dir::*;

pub mod file_mode;
