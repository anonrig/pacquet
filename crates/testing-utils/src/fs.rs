use std::{io, path::Path};

pub fn get_filenames_in_folder(path: &Path) -> Vec<String> {
    let mut files = std::fs::read_dir(path)
        .unwrap()
        .map(|e| e.unwrap().file_name().to_string_lossy().to_string())
        .collect::<Vec<_>>();

    files.sort();
    files
}

pub fn get_all_folders(root: &std::path::Path) -> Vec<String> {
    let mut files = Vec::new();
    for entry in walkdir::WalkDir::new(root) {
        let entry = entry.unwrap();
        let entry_path = entry.path();
        if entry.file_type().is_dir() || entry.file_type().is_symlink() {
            // We need this mutation to ensure that both Unix and Windows paths resolves the same.
            // TODO: Find a better way to do this?
            let simple_path = entry_path
                .strip_prefix(root)
                .unwrap()
                .components()
                .map(|c| c.as_os_str().to_str().expect("invalid UTF-8"))
                .collect::<Vec<_>>()
                .join("/");

            if !simple_path.is_empty() {
                files.push(simple_path);
            }
        }
    }
    files.sort();
    files
}

// Helper function to check if a path is a symlink or junction
pub fn is_symlink_or_junction(path: &Path) -> io::Result<bool> {
    #[cfg(windows)]
    return junction::exists(&path);

    #[cfg(not(windows))]
    return Ok(path.is_symlink());
}
