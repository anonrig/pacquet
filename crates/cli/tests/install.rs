use assert_cmd::prelude::*;
use command_extra::CommandExtra;
use pacquet_testing_utils::{
    bin::pacquet_with_temp_cwd,
    fs::{get_all_files, get_all_folders, is_symlink_or_junction},
};
use pipe_trait::Pipe;
use std::fs;

#[test]
fn should_install_dependencies() {
    let (command, root, workspace) = pacquet_with_temp_cwd(true);

    eprintln!("Creating package.json...");
    let manifest_path = workspace.join("package.json");
    let package_json_content = serde_json::json!({
        "dependencies": {
            "is-odd": "3.0.1",
        },
        "devDependencies": {
            "fast-decode-uri-component": "1.0.1",
        },
    });
    fs::write(&manifest_path, package_json_content.to_string()).expect("write to package.json");

    eprintln!("Executing command...");
    command.with_arg("install").assert().success();

    eprintln!("Make sure the package is installed");
    assert!(is_symlink_or_junction(&workspace.join("node_modules/is-odd")).unwrap());
    assert!(workspace.join("node_modules/.pacquet/is-odd@3.0.1").exists());

    eprintln!("Make sure it installs direct dependencies");
    assert!(!workspace.join("node_modules/is-number").exists());
    assert!(workspace.join("node_modules/.pacquet/is-number@6.0.0").exists());

    eprintln!("Make sure we install dev-dependencies as well");
    assert!(
        is_symlink_or_junction(&workspace.join("node_modules/fast-decode-uri-component")).unwrap()
    );
    assert!(workspace.join("node_modules/.pacquet/fast-decode-uri-component@1.0.1").is_dir());

    eprintln!("Snapshot");
    let workspace_folders = get_all_folders(&workspace);
    let store_files = get_all_files(&root.path().join("pacquet-store"));
    insta::assert_debug_snapshot!((workspace_folders, store_files));

    drop(root); // cleanup
}

#[test]
fn should_install_exec_files() {
    let (command, root, workspace) = pacquet_with_temp_cwd(true);

    eprintln!("Creating package.json...");
    let manifest_path = workspace.join("package.json");
    let package_json_content = serde_json::json!({
        "dependencies": {
            "pretty-exec": "0.3.10",
        },
    });
    fs::write(&manifest_path, package_json_content.to_string()).expect("write to package.json");

    eprintln!("Executing command...");
    command.with_arg("install").assert().success();

    eprintln!("Listing all files in the store...");
    let store_files = root.path().join("pacquet-store").pipe_as_ref(get_all_files);
    let (exec_files, non_exec_files): (Vec<_>, Vec<_>) =
        store_files.iter().partition(|path| path.ends_with("-exec"));

    eprintln!("Snapshot");
    insta::assert_debug_snapshot!(exec_files);

    #[cfg(unix)]
    {
        use pacquet_testing_utils::fs::is_path_executable;
        use pretty_assertions::assert_eq;

        eprintln!("All files that end with '-exec' are executable, others not");
        assert_eq!(
            store_files.iter().partition(|name| {
                root.path().join("pacquet-store").join(name).pipe_as_ref(is_path_executable)
            }),
            (exec_files, non_exec_files),
        );
    }

    #[cfg(windows)]
    let _ = non_exec_files;

    drop(root); // cleanup
}
