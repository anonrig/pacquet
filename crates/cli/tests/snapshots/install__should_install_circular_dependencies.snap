---
source: crates/cli/tests/install.rs
assertion_line: 155
expression: index_file_contents
---
v3/files/0d/920a93eb75934e4697e71f74e145832cba488947632806be7dd444871f575dae1acdf77ce06eab4653efa8b540d499baf374cc2783e153e68f68ec561dbff2-index.json:
  dependency.js:
    integrity: sha512-7/zrEcP4pbUxDUUXew2jxrJXssc/fT9Ioi+m/PpMBx7IOe1uTLaKdsPgdrSOdvfYVdayelgJyadaffSkeGpk+Q==
    mode: 420
    size: 59
  index.js:
    integrity: sha512-WyOuCJv5Gp2GVcdhsN759SYpgzGGCJSGSBnrN+vX1orAcy3PADNmOx2QYH13XUNtrML1klOSz/MuGOoPpKypjA==
    mode: 420
    size: 54
  mirror.js:
    integrity: sha512-xCuOztCcd9NCuMiCd2uz8gVggQnLPy6ck7cfCvzeM7ECofNMB1FoPYcADaOaquGH/3auJc/4U/xyJqDQyMpQIA==
    mode: 420
    size: 70
  package.json:
    integrity: sha512-lPjrGKy0dv4Mvtnp2BsqEYNxFGnOeHzpGvsRKK/UyQz4zGSXR4BoscLNMlK3mnVTt363x1tbIearrJH5DPBCFg==
    mode: 420
    size: 366
v3/files/9b/576e04418c866f0707ddb06734c8143ab825eaf283bae715282ad19acb540c27a8eb7f443cb3cb25c46043526824c11fd8d6516f27476ae9a0895f27af295b-index.json:
  dependency.js:
    integrity: sha512-As3lsS+ApUPQRjHQyIvNe0fHPXucVRl9eY2HJOLr6k/beGOnmdc+BOsj0woBa333z4jC6/3vuKwgvOo3UzX8mg==
    mode: 420
    size: 59
  index.js:
    integrity: sha512-WyOuCJv5Gp2GVcdhsN759SYpgzGGCJSGSBnrN+vX1orAcy3PADNmOx2QYH13XUNtrML1klOSz/MuGOoPpKypjA==
    mode: 420
    size: 54
  mirror.js:
    integrity: sha512-ytgA9GF21kB0lylGaZPPyxxe9OAuFfK9FoG9xzT9ICyBNruPFlYvi8Nd83XxnyYgu1TM+RodaELGnfmy9RwrUg==
    mode: 420
    size: 70
  package.json:
    integrity: sha512-vOF6vj7Excyek9Mf8e9hM5AeUrd0e02enYHPqY4Vfx7TyltI8QkoQDje0uSSfUt7T0LuXdd61pxDm/idaOsB5Q==
    mode: 420
    size: 366

